taichi-dev/taichi
# [ci] [windows] Add Dockerfile for Windows build and test (CPU)

This PR is our first step towards a fully containerized Windows build and test environment. Currently support CPU backend only, eventually we plan to cover all target backends on Windows. 

How to verify:
We take a two-steps approach: First build the `Dockerfile_Buildtools_2019` to install the essential windows buildtools including clang, cmake, ninja etc. Then, based on the resulting image, we continue to add git, conda etc for Taichi build, by using `Dockerfile_cpu`. Finally, use `win_build_test` to download the prebuilt, clone Taichi, and build and test the latest master. All tests should pass. Steps are summarize as follows:

First build the image and start the container: 
```
docker build -f Dockerfile_Buildtools_2019 -t taichi-buildtools-2019:v0.0.1 -m 8GB .
docker build -f Dockerfile_cpu -t taichi-dev-windows:v0.0.1 -m 8GB.
docker run -it taichi-dev-windows:v0.0.1
```  
Note that `-m 8GB` is required, if you dont have 8GB, at least 6GB should be used. Otherwise, buildtools installation may fail silently (check the referenced issue in dockerfile).

After starting the container, you will be entered into the x64 VS Developer Command Prompt, you should observe `Environment initialized for: 'x64'`, otherwise the image is not properly built.

To run the build and test, in the command prompt: simply execute `win_build_test.ps1`, since we have already copied the script into the container. In the end, the tests should be finished with all passed.

Note that the current Dockerfile is not optimized in any ways (size, build time, etc). This PR serves as a proof of concept. Further optimization steps will be followed. Your suggestions and comments are welcome!
